### PR TITLE
[mlir][EmitC] Unify handling of operations which are emitted in a deferred way

### PR DESCRIPTION
**Pull request by simon-camp**
Several operations from the EmitC dialect don't produce output directly during emission, but rather when being used as an operand. These changes unify the handling of such operations and fix a bug in the emission of global ops.

Co-authored-by: Marius Brehler <marius.brehler@iml.fraunhofer.de>

### CODE REVIEW

**Review by aniragil — CHANGES_REQUESTED** [review]
Patch missing a more detailed description?

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp [hunk 1]
```diff
@@ -1001,7 +993,8 @@ static LogicalResult printFunctionBody(CppEmitter &emitter,
       // trailing semicolon is handled within the printOperation function.
       bool trailingSemicolon =
           !isa<cf::CondBranchOp, emitc::DeclareFuncOp, emitc::ForOp,
-               emitc::IfOp, emitc::LiteralOp, emitc::VerbatimOp>(op);
+               emitc::IfOp, emitc::VerbatimOp>(op) ||
+          hasDeferredEmission(&op);
```
[the remark this answers is not in the document]
Comment (simon-camp) [reply mlir/lib/Target/Cpp/TranslateToCpp.cpp]: Yes the check looks redundant to me here, I removed it.

**Review by aniragil — APPROVED** [review]
LGTM!

Is the bug fixed the one pointed out [here](https://discourse.llvm.org/t/mlir-emitc-translate-to-c/79647)?

File: mlir/lib/Target/Cpp/TranslateToCpp.cpp [hunk 2]
```diff
@@ -295,10 +304,10 @@ static bool shouldBeInlined(ExpressionOp expressionOp) {
 
   Operation *user = *result.getUsers().begin();
 
-  // Do not inline expressions used by subscript operations, since the
-  // way the subscript operation translation is implemented requires that
-  // variables be materialized.
-  if (isa<emitc::SubscriptOp>(user))
+  // Do not inline expressions used by operations with deferred emission, since
+  // the way their translation is implemented requires that variables be
+  // materialized.
```
Comment (marbre) [inline mlir/lib/Target/Cpp/TranslateToCpp.cpp:309]: Maybe rephrase to
```suggestion
  // the way their translation is implemented requires that variables are
  // materialized.
```
or
```suggestion
  // the way their translation is implemented requires to materialize
  // variables.
```
or
```suggestion
  // their translation requires the materialization of variables.
```

**Review by marbre — APPROVED** [review]
Looks good, just one minor suggestion.
